### PR TITLE
Application Structural Refactor

### LINKED ISSUE #19: Improve code structure
`Express` is a decent library/framework (who cares, really?), but it does not enforce any structure at all, which can result in getting stuff all over the place.

So, we should follow something like [this](https://softwareontheroad.com/ideal-nodejs-project-structure/) - I'm reading it now and it all makes a lot of sense. We can iterate on it, but I feel that at least a basic level of what's described there should be implemented :smile:

### LINKED ISSUE #25: Switch tests from Mocha+Chai to Jest
I now believe that Jest tests are much cleaner. Plus, Jest can run tests in parallel for better performance and has mocking built-in, as well as coverage reports!

### PR DESCRIPTION
**Pull request by miguelpduarte**
Closes #19. Closes #25.

Currently going for a mix of the Bulletproof architecture mentioned in the issue, Ruby on Rails (it has a pretty good architecture imo, despite ruby being weird) and some of my own opinions.

This is in the very early stages, but opinions are welcome as this is a core concern for future development.

### CODE REVIEW

**Comment by miguelpduarte** [comment]
Also closes #25

File: .env [hunk 1]
```diff
@@ -13,4 +13,4 @@ DB_NAME=nijobs
 SESSION_SECRET=O Rui foi membro do IEEE.
 
 # Specifies the port in which the app will be exposed
-PORT=8000
+PORT=8087
```
Comment (imnotteixeira) [inline .env]: Why this change? Anything in particular?

File: src/loaders/express.js [hunk 2]
```diff
@@ -23,33 +25,35 @@ module.exports = (app) => {
...
+    // Adds route logging
+    app.use(morgan("common"));
 
-        // Request methods to allow
+    // Adding headers (CORS)
+    app.use((_, res, next) => {
+        // Allow connections for all origins
+        res.setHeader("Access-Control-Allow-Origin", "*");
+        // Allowed request methods
         res.setHeader("Access-Control-Allow-Methods", "GET, POST, OPTIONS, PUT, PATCH, DELETE");
```
Comment (imnotteixeira) [inline src/loaders/express.js]: Are we using all of these? Shouldn't we remove the ones we don't use?

File: src/loaders/express.js [hunk 3]
```diff
@@ -23,33 +25,35 @@ module.exports = (app) => {
...
-    // Website to allow to connect
-        res.setHeader("Access-Control-Allow-Origin", "*");
+    // Adds route logging
+    app.use(morgan("common"));
 
-        // Request methods to allow
+    // Adding headers (CORS)
+    app.use((_, res, next) => {
+        // Allow connections for all origins
+        res.setHeader("Access-Control-Allow-Origin", "*");
```
Comment (imnotteixeira) [inline src/loaders/express.js]: We should have an `env` variable with the allowed origins, for more safety. It is not that hard and we could customize it to allow only our applications for now (nijobs-fe, for example). Surely for development environment, we could have the `*` the same way.

File: .env (hunk 1 above)
Comment (miguelpduarte) [reply .env]: Collided with a lot of stuff that opens listens on 8000 by default.

File: src/loaders/express.js (hunk 2 above)
Comment (miguelpduarte) [reply src/loaders/express.js]: This is CORS, which will be necessary for any API.

File: .env (hunk 1 above)
Comment (imnotteixeira) [reply .env]: :+1:

File: src/loaders/express.js (hunk 2 above)
Comment (miguelpduarte) [reply src/loaders/express.js]: If you are mentioning the allowed methods, this was probably just a left-over copy pasta from somewhere online for enabling CORS for node+express

File: src/loaders/express.js (hunk 3 above)
Comment (miguelpduarte) [reply src/loaders/express.js]: This is a valid future concern, please create an issue mentioning your comment for increased context.

File: src/loaders/express.js (hunk 2 above)
Comment (imnotteixeira) [reply src/loaders/express.js]: I was asking if all the http methods there will be used, if not, they can be removed from the `Access-Control-Allow-Methods` Header

Comment (miguelpduarte) [reply src/loaders/express.js]: I don't really see a problem, as they would not match any endpoint and just return a 404. But I get what you mean.

**Comment by miguelpduarte** [comment]
Derped with the force pushes, fixed the problem and amended the commit but forgot to add the files 🙃 

Sorry for the spam there 😅

File: src/loaders/express.js (hunk 3 above)
Comment (imnotteixeira) [reply src/loaders/express.js]: Here you go: #41

Comment (miguelpduarte) [reply src/loaders/express.js]: Thanks!

**Comment by miguelpduarte** [comment]
The tests were hanging, and jest traced it to some `supertest` stuff. Found https://github.com/nodkz/mongodb-memory-server and in fact it seems that the problem was not disconnecting from the `mongo` server in the end of the test suites 🙃

File: src/api/middleware/errorHandler.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
+const { isCelebrate } = require("celebrate");
+const EscapeHtml = require("escape-html");
+
+const { errorExtractor } = require("../../lib/dbErrorExtractor");
+
+const ERROR_TYPES = Object.freeze({
```
Comment (imnotteixeira) [inline src/api/middleware/errorHandler.js]: Let's rename this variable to ErrorTypes, so it can be used as ErrorTypes.FORBIDDEN, for example, I think it is more common, at least in java. (does not have to be changed here, but it should, to ensure consistency)

File: src/lib/dbErrorExtractor.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    UNDETERMINED_ERROR: 1,
+    DUPLICATED_KEY: 2,
+});
+
+const errorExtractor = (err) => {
+    if (handlers.hasOwnProperty(err.code)) {
+        return handlers[err.code](err);
+    }
+    return {
+        type: DB_ERROR_TYPES,
```
Comment (imnotteixeira) [inline src/lib/dbErrorExtractor.js]: This should be an actual error type, not the enum :D

File: src/lib/dbErrorExtractor.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
+const DB_ERROR_TYPES = Object.freeze({
```
Comment (imnotteixeira) [inline src/lib/dbErrorExtractor.js]: Same thing, change to `DBErrorTypes`

File: src/loaders/express.js [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        res.setHeader("Access-Control-Allow-Headers", "X-Requested-With, content-type, authorization");
+        // Because we need the website to include cookies in the requests sent
+        // to the API (we are using sessions)
+        res.setHeader("Access-Control-Allow-Credentials", true);
+        // Continue to next layer of middleware
+        return next();
+    });
+
+    // Health check endpoint
+    app.get("/", (_, res) => res.status(200).json({ "online": true }));
```
Comment (imnotteixeira) [inline src/loaders/express.js]: The fact that the server is answering (200 response) is enough to know that it is online. Instead, we could send the API version in the response (add it to .env?)

File: src/services/auth.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
+// TODO: Use typedi or similar
+const Account = require("../models/Account");
+
+class AuthService {
+    constructor() {
```
Comment (imnotteixeira) [inline src/services/auth.js]: Empty constructor is not required

File: src/services/auth.js [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+    }
+
+    async register(username, password) {
+        const account = await Account.create({
+            username: username,
```
Comment (imnotteixeira) [inline src/services/auth.js]: You can use the smaller syntax 
```js
{
    username,
    password
}
```

File: src/services/auth.js [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+            password: password,
+        });
+
+        return {
+            username: account.username,
```
Comment (imnotteixeira) [inline src/services/auth.js]: We can use `{ username }` directly, since it will be the same as `account.username`

File: src/setupTests.js [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+//----------------------------------
+
+// During the test the env variable must be set to test
+if (process.env.NODE_ENV !== "test") {
+    console.error("Entered test files without being in the test environment, aborting!");
+    process.exit(55);
+}
+
+const mongoose = require("mongoose");
+// Ocurrs before all the tests, only once
```
Comment (imnotteixeira) [inline src/setupTests.js]: Grammarly your code: `occurs`

File: src/api/middleware/errorHandler.js [hunk 12]
```diff
@@ -32,10 +34,23 @@ const handler = () => (err, req, res, next) => {
...
+        return next(err);
+    }
+
+    const result = {
+        success: false,
+        error_code: ERROR_TYPES.DB_ERROR,
+        error: errorExtractor(err),
+    };
+
+    return res.status(500).send(result);
```
Comment (miguelpduarte) [inline src/api/middleware/errorHandler.js]: Currently, in the only case that this happens the status should probably be 400. But we should revisit this later, the status should be calculated from some more context, such as a `req.something` added by the request handler to give context to the db error handler.

File: src/api/middleware/validators/auth.js [hunk 13]
```diff
@@ -0,0 +1,10 @@
+const { celebrate, Joi } = require("celebrate");
```
Comment (miguelpduarte) [inline src/api/middleware/validators/auth.js]: We should switch to [`express-validator`](https://www.npmjs.com/package/express-validator) because of this: https://www.npmtrends.com/celebrate-vs-express-validator

File: src/loaders/express.js (hunk 7 above)
Comment (miguelpduarte) [reply src/loaders/express.js]: The json content could be empty, this content is just so that it is not, for easier checking of the app being online via a browser or something that might hide an empty response.

Comment (miguelpduarte) [reply src/loaders/express.js]: I don't think we have discussed versioning, so I'd rather not dive into that now (at least not in this PR, it's already huge)

File: src/services/auth.js (hunk 8 above)
Comment (miguelpduarte) [reply src/services/auth.js]: This had the purpose of setting up dependency injection using something like `typedi`, but was stopped in the interim due to other things being more relevant at the time.

Comment (miguelpduarte) [reply src/services/auth.js]: However, it can be removed without any issue

File: src/services/auth.js (hunk 10 above)
Comment (miguelpduarte) [reply src/services/auth.js]: Ideally so, but some hooks in the Model might change it, due to trimming or something of the sort. I see no problem with leaving this as is.

File: src/setupTests.js (hunk 11 above)
Comment (miguelpduarte) [reply src/setupTests.js]: :cry:

**Comment by miguelpduarte** [comment]
Resolved the conflicts via rebasing develop into this branch

File: src/services/auth.js (hunk 9 above)
Comment (imnotteixeira) [reply src/services/auth.js]: bump

Comment (miguelpduarte) [reply src/services/auth.js]: A bit nitpicking, but fixed :stuck_out_tongue:
